### PR TITLE
Adding endpoint to fetch logs by runid

### PR DESCRIPTION
**Pull request by poornachandra**
JIRA https://issues.cask.co/browse/CDAP-2051

Adding log endpoints to fetch logs by run-id -
* `/namespaces/{namespace-id}/apps/{app-id}/{program-type}/{program-id}/runs/{run-id}/logs`
* `/namespaces/{namespace-id}/apps/{app-id}/{program-type}/{program-id}/runs/{run-id}/logs/next`
* `/namespaces/{namespace-id}/apps/{app-id}/{program-type}/{program-id}/runs/{run-id}/logs/prev`
* `/namespaces/{namespace-id}/adapters/{adapter-id}/runs/{run-id}/logs`

Also refactored test cases.

### CODE REVIEW

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/context/LoggingContextHelper.java [hunk 1]
```diff
@@ -189,6 +201,7 @@ public static Filter createFilter(LoggingContext loggingContext) {
                                                                  NamespaceLoggingContext.TAG_NAMESPACE_ID, namespaceId),
                                                                new MdcExpression(ACCOUNT_ID,
                                                                                  Constants.DEVELOPER_ACCOUNT)));
+      // Filter for Adapters
       if (loggingContext.getSystemTagsMap().containsKey(ApplicationLoggingContext.TAG_ADAPTER_ID)) {
```
Comment (bdmogal) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/context/LoggingContextHelper.java]: shouldn't there be a run id here?

File: cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/log/MockLogReader.java [hunk 2]
```diff
@@ -115,6 +119,10 @@ public void getLogNext(LoggingContext loggingContext, LogOffset fromOffset, int
             continue;
           }
 
+          if (runId != null && !logLine.getRunId().equals(runId)) {
+            continue;
+          }
+
           callback.handle(
             new LogEvent(
               new LoggingEvent("com.continuiity.Test",
```
Comment (bdmogal) [inline cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/log/MockLogReader.java]: nit: should we change these?

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/context/LoggingContextHelper.java [hunk 3]
```diff
@@ -127,19 +127,31 @@ public static LoggingContext getLoggingContext(String namespaceId, String applic
 
   public static LoggingContext getLoggingContext(String namespaceId, String applicationId, String entityId,
                                                  ProgramType programType, @Nullable String adapterName) {
+    return getLoggingContext(namespaceId, applicationId, entityId, programType, null, adapterName);
+  }
+
+  public static LoggingContext getLoggingContextWithRunId(String namespaceId, String applicationId, String entityId,
```
Comment (bdmogal) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/context/LoggingContextHelper.java]: Is this method necessary? It only enforces a runId (via the absence of ``@Nullable``, right? can we directly call the method below?

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/gateway/handlers/LogHandler.java [hunk 4]
```diff
@@ -156,42 +218,46 @@ public void prev(HttpRequest request, HttpResponder responder, @PathParam("names
...
-      responder.sendStatus(HttpResponseStatus.UNAUTHORIZED);
-    } catch (IllegalArgumentException e) {
-      responder.sendString(HttpResponseStatus.BAD_REQUEST, e.getMessage());
-    } catch (Throwable e) {
-      LOG.error("Caught exception", e);
-      responder.sendStatus(HttpResponseStatus.INTERNAL_SERVER_ERROR);
-    }
+  @GET
+  @Path("/namespaces/{namespace-id}/adapters/{adapter-id}/runs/{run-id}/logs")
+  public void getAdapterRunIdLogs(HttpRequest request, HttpResponder responder,
```
Comment (bdmogal) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/gateway/handlers/LogHandler.java]: So adapter next and prev aren't part of this PR right? Can we report a jira to track?

File: cdap-watchdog/src/test/java/co/cask/cdap/logging/appender/LoggingTester.java [hunk 5]
```diff
@@ -16,23 +16,98 @@
...
+      replaceTag(loggingContextNs2, new Entry(ApplicationLoggingContext.TAG_RUNID_ID, null),
+                 new Entry(ApplicationLoggingContext.TAG_INSTANCE_ID, null)));
+    for (int i = 80; i < 120; ++i) {
+      logger.warn("NS_2 Test log message {} {} {}", i, "arg1", "arg2", e2);
+    }
+
+    ByteArrayOutputStream bos = new ByteArrayOutputStream();
+    StatusPrinter.setPrintStream(new PrintStream(bos));
+    StatusPrinter.print((LoggerContext) LoggerFactory.getILoggerFactory());
+    System.out.println(bos.toString());
```
Comment (bdmogal) [inline cdap-watchdog/src/test/java/co/cask/cdap/logging/appender/LoggingTester.java]: stdout? Is this intentional?

**Comment by bdmogal** [comment]
@poornachandra mostly looks good. Just a few minor comments. Only one major about runId in adapters.

File: cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterPathLookup.java [hunk 6]
```diff
@@ -155,10 +155,13 @@ private String getV3RoutingService(String [] uriParts, AllowedMethod requestMeth
         return Constants.Service.STREAMS;
       }
     } else if ((uriParts.length >= 8 && uriParts[7].equals("logs")) ||
+      (uriParts.length >= 10 && uriParts[9].equals("logs")) ||
```
Comment (chtyim) [inline cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterPathLookup.java]: I "love" this router code..... is there a way to do better than this, like by prefix only? Maybe our v3 API is not good?

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/context/LoggingContextHelper.java (hunk 1 above)
Comment (poornachandra) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/context/LoggingContextHelper.java]: Good point, I missed it here.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/context/LoggingContextHelper.java (hunk 3 above)
Comment (poornachandra) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/context/LoggingContextHelper.java]: Due to having two `@Nullable`s it becomes confusing to call the other method.

File: cdap-watchdog/src/test/java/co/cask/cdap/logging/appender/LoggingTester.java (hunk 5 above)
Comment (poornachandra) [inline cdap-watchdog/src/test/java/co/cask/cdap/logging/appender/LoggingTester.java]: Yes, logger errors cannot be reported using logging system.

File: cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterPathLookup.java (hunk 6 above)
Comment (poornachandra) [inline cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterPathLookup.java]: Some URLs cannot be prefix-matched. I think we need to do routing by getting all the URLs that need to be routed, and pattern match on it.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/gateway/handlers/LogHandler.java (hunk 4 above)
Comment (poornachandra) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/gateway/handlers/LogHandler.java]: Filed JIRA - https://issues.cask.co/browse/CDAP-2107

**Comment by poornachandra** [comment]
@bdmogal Addressed review comments, please take a look.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/context/LoggingContextHelper.java [hunk 7]
```diff
@@ -189,21 +181,60 @@ public static Filter createFilter(LoggingContext loggingContext) {
...
+      entityId = loggingContext.getSystemTagsMap().get(tagName).getValue();
+    } else if (loggingContext instanceof SparkLoggingContext) {
+      tagName = SparkLoggingContext.TAG_SPARK_JOB_ID;
+      entityId = loggingContext.getSystemTagsMap().get(tagName).getValue();
+    } else if (loggingContext instanceof UserServiceLoggingContext) {
+      tagName = UserServiceLoggingContext.TAG_USERSERVICE_ID;
+      entityId = loggingContext.getSystemTagsMap().get(tagName).getValue();
+    } else if (loggingContext instanceof WorkerLoggingContext) {
+      tagName = WorkerLoggingContext.TAG_WORKER_ID;
+      entityId = loggingContext.getSystemTagsMap().get(tagName).getValue();
```
Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/context/LoggingContextHelper.java]: ``entityId = loggingContext.getSystemTagsMap().get(tagName).getValue();`` is common in all conditions, may be we can move that outside.

Comment (poornachandra) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/context/LoggingContextHelper.java]: Good point, I missed it.

**Comment by bdmogal** [comment]
:+1: after @gsps1's comment

**Comment by poornachandra** [comment]
@gsps1 Addressed comments, please take a look.

**Comment by poornachandra** [comment]
Build - http://builds.cask.co/browse/CDAP-DUT1429-9

**Pull request merged**